### PR TITLE
feat: add `check_connected_accounts` flag on `get_tools`

### PR DESCRIPTION
**Pull request by tushar-composio**
> [!IMPORTANT]
> Add `check_connected_accounts` flag to `ComposioToolSet` to control connected account validation, with tests and warnings if disabled.
> 
>   - **Behavior**:
>     - Add `check_connected_accounts` flag to `ComposioToolSet` constructor to control connected account validation.
>     - If `check_connected_accounts` is `False`, skips validation in `_validating_connection_ids()` and `get_action_schemas()`.
>     - Warns if `check_connected_accounts` is `False` in `get_action_schemas()`.
>   - **Tests**:
>     - Add `test_check_connected_accounts_flag` in `test_toolset.py` to verify behavior of `check_connected_accounts` flag.
>   - **Misc**:
>     - Minor formatting changes in `_serialize_execute_params()` in `toolset.py`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=SamparkAI%2Fcomposio&utm_source=github&utm_medium=referral)<sup> for 0a8ce00dfd91e6f0916a29c12643ade5d2395657. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: python/tests/test_tools/test_toolset.py [hunk 1]
```diff
@@ -240,3 +240,19 @@ def postprocess(response: dict) -> dict:
     toolset.get_tools(processors={"post": {Action.COMPOSIO_LIST_APPS: postprocess}})
     toolset.execute_action(Action.COMPOSIO_LIST_APPS, {})
     assert postprocessor_called
+
+
+def test_check_connected_accounts_flag(monkeypatch: pytest.MonkeyPatch) -> None:
```
[the remark this answers is not in the document]
Comment (tushar-composio) [reply python/tests/test_tools/test_toolset.py:245]: LangchainToolSet is a subclass that uses the same constructor, this is a better test than importing the base toolset directly.

**Review by angrybayblade — APPROVED** [review]

**Comment by shreysingla11** [comment]
Great addition of the `check_connected_accounts` parameter! To further improve the code, consider adding a more detailed docstring for this parameter in the `get_action_schemas` method. This would help developers understand when and why they might want to use this option. For example:

```python
def get_action_schemas(
    self,
    apps: t.Optional[t.Sequence[AppType]] = None,
    actions: t.Optional[t.Sequence[ActionType]] = None,
    tags: t.Optional[t.Sequence[TagType]] = None,
    *,
    check_connected_accounts: bool = True,
) -> t.List[ActionModel]:
    """
    Get action schemas for the specified apps, actions, or tags.

    :param check_connected_accounts: If True (default), verifies connected accounts for each action.
                                     Set to False to skip verification, which can be useful in scenarios
                                     where account verification is not necessary or has been done elsewhere.
                                     Note: Setting this to False may lead to runtime errors if accounts
                                     are not properly configured when actions are executed.
    """
    # ... rest of the method implementation
```

This provides more context about the parameter's purpose and potential risks, which will be valuable for future developers working with this code.

**Comment by shreysingla11** [comment]
Overall, this PR introduces a valuable feature that adds flexibility to the `ComposioToolSet` class and its derivatives. Here's a summary of my review:

1. Consistency: The implementation is consistent across different plugins, which is excellent for maintainability.

2. Non-breaking change: The default behavior (checking connected accounts) is maintained, ensuring backwards compatibility.

3. Warning message: The addition of a warning when `check_connected_accounts` is False is a good practice to prevent potential runtime errors.

4. Test coverage: The new test in `test_toolset.py` is a good addition to verify the behavior of the new flag.

Suggestions for further improvement:

1. Consider adding logging statements when account verification is skipped, in addition to the warning. This could aid in debugging.

2. Ensure that all plugin-specific implementations handle this new parameter consistently.

3. You might want to add a brief comment in the plugin-specific `get_tools` methods to explain why the `check_connected_accounts` parameter is being passed through to `get_action_schemas`.

Overall, this is a well-implemented feature that enhances the flexibility of the `ComposioToolSet` while maintaining good coding practices. Great job!

File: python/plugins/camel/composio_camel/toolset.py [hunk 2]
```diff
@@ -180,5 +181,10 @@ def get_tools(
                 ).model_dump(),
                 entity_id=entity_id,
             )
```
Comment (shreysingla11) [inline python/plugins/camel/composio_camel/toolset.py:183]: The `check_connected_accounts` parameter is added to the `get_tools()` method, but it's not being passed to the `get_action_schemas()` method where the actual account checking occurs. Consider updating the `get_tools()` method to pass this parameter:

```python
def get_tools(
    self,
    actions: t.Optional[t.Sequence[ActionType]] = None,
    apps: t.Optional[t.Sequence[AppType]] = None,
    tags: t.Optional[t.List[TagType]] = None,
    entity_id: t.Optional[str] = None,
    *,
    processors: t.Optional[ProcessorsType] = None,
    check_connected_accounts: bool = True,
) -> t.List[OpenAIFunction]:
    # ... existing code ...
    return [
        self._wrap_tool(
            # ... existing code ...
        )
        for schema in self.get_action_schemas(
            actions=actions,
            apps=apps,
            tags=tags,
            check_connected_accounts=check_connected_accounts
        )
    ]
```

This change ensures that the `check_connected_accounts` flag is properly utilized.

File: python/composio/tools/toolset.py [hunk 3]
```diff
@@ -809,8 +749,15 @@ def get_action_schemas(
                 actions=remote_actions,
                 tags=tags,
             )
-            for item in remote_items:
-                self.check_connected_account(action=item.name)
+            if self._check_connected_accounts:
+                for item in remote_items:
+                    self.check_connected_account(action=item.name)
+            else:
+                warnings.warn(
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:818]: The implementation of the `check_connected_accounts` flag looks good here. However, consider adding a suggestion to the user on how to ensure accounts are connected if needed. For example:

```python
warnings.warn(
    "Not verifying connected accounts for apps. "
    "Actions may fail when the Agent tries to use them. "
    "Ensure all required accounts are connected before using the tools.",
    UserWarning,
)
```

This additional information can help users understand the implications of setting `check_connected_accounts` to `False` and how to mitigate potential issues.

File: python/composio/tools/toolset.py [hunk 4]
```diff
@@ -158,6 +148,8 @@ def __init__(
         output_dir: t.Optional[Path] = None,
         verbosity_level: t.Optional[int] = None,
         connected_account_ids: t.Optional[t.Dict[AppType, str]] = None,
+        *,  # TODO: move all keyword args below this `*` in a future release
+        check_connected_accounts: bool = True,
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py]: Consider adding a docstring for the `check_connected_accounts` parameter to explain its purpose and default value. For example:

```python
def __init__(
    self,
    # ... other parameters ...
    *,
    check_connected_accounts: bool = True,
    **kwargs: t.Any,
) -> None:
    """
    Initialize the ComposioToolSet.

    :param check_connected_accounts: If True, verify connected accounts for each action.
                                     Set to False to skip verification, which may improve
                                     performance but could lead to runtime errors if
                                     accounts are not properly connected.
    """
    # ... rest of the method ...
```

This documentation will help users understand the purpose and implications of the `check_connected_accounts` parameter.

**Comment by shreysingla11** [comment]
Thank you for implementing the `check_connected_accounts` flag. Here's a summary of my review:

1. The implementation in `get_action_schemas()` looks good, but consider enhancing the warning message as suggested in the inline comment.

2. Add a docstring for the `check_connected_accounts` parameter in the `__init__` method to explain its purpose and implications.

3. Ensure that all toolset implementations (camel, claude, griptape, etc.) are updated consistently with this change. Each `get_tools()` method should pass the `check_connected_accounts` flag to `get_action_schemas()`.

4. Consider updating existing tests or adding new ones to cover the behavior of the `check_connected_accounts` flag, including cases where it's set to `False`.

5. If there are any performance implications of setting `check_connected_accounts` to `True` or `False`, consider adding a note in the documentation.

Overall, this is a good addition that provides more flexibility to users. Once these minor improvements are made, the PR should be ready for merge. Great work!

**Comment by shreysingla11** [comment]
Thank you for this pull request adding the `check_connected_accounts` flag to the `get_tools()` method. This feature adds valuable flexibility to the toolset usage. Here's a summary of the analysis and suggestions for improvement:

1. Documentation:
   - Consider adding a docstring for the `check_connected_accounts` parameter in the `get_action_schemas()` method to explain its purpose and potential risks.
   - Update the user guide or API documentation to clearly explain the implications of setting `check_connected_accounts=False`.

2. Error Handling:
   - Consider adding a try-except block around the `self.check_connected_account(action=item.name)` calls to handle potential exceptions gracefully.

3. Testing:
   - The test added for the new feature is well-structured. Consider expanding test coverage to include more scenarios, edge cases, and different combinations of actions and apps.

4. Performance Consideration:
   - If the `check_connected_account` method is time-consuming, consider implementing a caching mechanism to avoid repeated checks for the same action within a session.

5. Consistency:
   - Ensure that all plugin-specific toolsets handle the `check_connected_accounts` flag consistently. Double-check that no toolset was missed in this update.

Specific file comments:

1. In `python/composio/tools/toolset.py`, around line 777:
   Consider adding a docstring for the `check_connected_accounts` parameter to explain its purpose and potential risks.

2. In `python/composio/tools/toolset.py`, around line 815:
   It might be beneficial to add a try-except block here to handle potential exceptions from `check_connected_account` gracefully.

3. In `python/tests/test_tools/test_toolset.py`, around line 248:
   Good job on adding tests for the new feature. Consider adding more test cases for different combinations of actions and apps, and potential edge cases.

Overall, this change adds useful flexibility to the toolset. The implementation is solid, with room for minor improvements in documentation and error handling. Great work on maintaining code style and adding appropriate tests! Just ensure that the potential risks of skipping account verification are well-documented for users.

**Comment by shreysingla11** [comment]
Final Review Summary for PR #746:

Overall, the implementation of the `check_connected_accounts` flag is well-executed and consistent across the codebase. Here's a summary of the findings:

1. Consistency: The `check_connected_accounts` parameter has been correctly added to the `get_tools()` method across all relevant plugin files (Camel, Claude, Griptape, Langchain, LlamaIndex, Lyzr, OpenAI, Phidata, and PraisonAI). This ensures a uniform interface across different toolsets.

2. Functionality: The flag is properly used in the `get_action_schemas()` method of the base `ComposioToolSet` class, with appropriate logic to skip account verification when set to False.

3. Warning Implementation: A warning is correctly issued when `check_connected_accounts` is set to False, alerting users to potential risks.

4. Testing: A test case has been added in `test_toolset.py` to verify the behavior of the new flag, which is a positive addition for ensuring functionality.

Suggestions for further improvement:

1. Documentation: Consider adding docstrings for the `check_connected_accounts` parameter in the `get_action_schemas()` method and in each plugin's `get_tools()` method. This would improve code clarity and help users understand the purpose and implications of this flag.

2. Error Handling: The previously suggested try-except block around `self.check_connected_account(action=item.name)` has not been implemented. This could be considered for future improvements to handle potential exceptions more gracefully.

3. Expanded Testing: While a test has been added, more comprehensive tests covering different scenarios (e.g., with various combinations of actions and apps) could further strengthen the implementation.

4. Performance Consideration: If the `check_connected_account` method proves to be time-consuming in practice, implementing a caching mechanism could be considered in the future to optimize performance.

5. User Guide Update: Ensure that the user guide or API documentation is updated to explain the new `check_connected_accounts` flag, its default behavior, and the implications of setting it to False.

In conclusion, the implementation is solid and ready for merge, pending any final documentation updates. The consistent implementation across all toolsets is particularly commendable. The suggested improvements are minor and could be addressed in future iterations if deemed necessary.

Great work on this feature addition!

**Comment by shreysingla11** [comment]
Review Conclusion:

Based on our thorough analysis of the changes implemented in this pull request, we consider this review complete. The implementation of the `check_connected_accounts` flag is consistent and well-integrated across the codebase.

Key points:
1. The feature has been correctly implemented across all relevant files.
2. A test case has been added to verify the new functionality.
3. The changes align well with the existing codebase structure and conventions.

While we've suggested some minor improvements for future consideration (such as additional documentation and expanded test coverage), these are not blocking issues for the current implementation.

Unless there are any objections or additional changes requested by the development team, this pull request appears ready for final approval and merge.

Thank you for the opportunity to review these changes. Great work on this feature addition!

File: python/plugins/camel/composio_camel/toolset.py [hunk 5]
```diff
@@ -154,6 +154,7 @@ def get_tools(
         entity_id: t.Optional[str] = None,
         *,
         processors: t.Optional[ProcessorsType] = None,
+        check_connected_accounts: bool = True,
```
Comment (shreysingla11) [inline python/plugins/camel/composio_camel/toolset.py:157]: The `check_connected_accounts` flag has been added to the `get_tools()` method, which is a good addition for flexibility. However, it might be beneficial to add a brief comment explaining the purpose of this flag where it's introduced in each file for better readability and maintainability.

File: python/plugins/camel/composio_camel/toolset.py [hunk 6]
```diff
@@ -154,6 +154,7 @@ def get_tools(
         entity_id: t.Optional[str] = None,
```
Comment (shreysingla11) [inline python/plugins/camel/composio_camel/toolset.py:154]: Consider updating the docstring for the `get_tools()` method to include information about the new `check_connected_accounts` parameter, its default value, and its purpose. This will help users understand how to use this new feature.

File: python/composio/tools/toolset.py [hunk 7]
```diff
@@ -809,8 +811,15 @@ def get_action_schemas(
                 actions=remote_actions,
                 tags=tags,
             )
-            for item in remote_items:
-                self.check_connected_account(action=item.name)
+            if check_connected_accounts:
+                for item in remote_items:
+                    self.check_connected_account(action=item.name)
+            else:
+                warnings.warn(
```
Comment (shreysingla11) [inline python/composio/tools/toolset.py:818]: It's good that a warning is issued when `check_connected_accounts` is set to `False`. However, consider adding more specific information in the warning message about potential consequences or scenarios where this might be useful. This will help users make informed decisions when using this flag.

**Comment by shreysingla11** [comment]
Overall, the addition of the `check_connected_accounts` flag to the `get_tools()` method is a good improvement that adds flexibility to the toolset. The implementation is consistent across different plugins, which is commendable. Here's a summary of the review:

1. The new flag is properly implemented in various toolset classes.
2. A warning is issued when the flag is set to `False`, which is a good practice.
3. The changes are consistent across different plugin implementations.

Suggestions for improvement:
1. Add brief comments explaining the purpose of the `check_connected_accounts` flag where it's introduced in each file.
2. Update the docstrings for the `get_tools()` method to include information about the new parameter.
3. Consider providing more detailed information in the warning message about potential consequences or use cases for setting the flag to `False`.

These changes will enhance code readability, improve documentation, and help users make informed decisions when using this new feature. Once these minor improvements are addressed, the PR should be ready for merging.
